File: web/src/utils/analysis.js
import { jStat } from 'jstat';

/**
 * Convert absorbance to melanin concentration using standard curve.
 * Conc = (Abs - intercept) / slope
 */
export function absorbanceToConcentration(absorbance, slope, intercept) {
  if (slope === 0) throw new Error('Slope cannot be zero.');
  return (absorbance - intercept) / slope;
}

/**
 * Calculate melanin concentrations for each group across all measurements.
 *
 * For each file, for each group: average the absorbance of assigned wells,
 * then convert to concentration via standard curve.
 *
 * @returns {{ [groupName]: number[] }} concentration per measurement
 */
export function calculateGroupConcentrations(parsedData, conditions, slope, intercept) {
  const result = {};
  for (const cond of conditions) {
    result[cond.name] = [];
  }

  const files = Object.values(parsedData);
  for (const fileData of files) {
    for (const cond of conditions) {
      const absValues = cond.wells
        .filter((w) => w in fileData)
        .map((w) => fileData[w]);

      if (absValues.length > 0) {
        const meanAbs = absValues.reduce((a, b) => a + b, 0) / absValues.length;
        const conc = absorbanceToConcentration(meanAbs, slope, intercept);
        result[cond.name].push(conc);
      }
    }
  }

  return result;
}

/**
 * Normalize all concentrations to the mean of the reference group.
 * Each value becomes (value / refMean) * 100.
 */
export function normalizeToReference(concentrations, referenceGroup) {
  const refValues = concentrations[referenceGroup];
  if (!refValues || refValues.length === 0) {
    throw new Error(`Reference group '${referenceGroup}' has no data.`);
  }

  const refMean = refValues.reduce((a, b) => a + b, 0) / refValues.length;
  if (refMean === 0) {
    throw new Error('Reference group mean is zero; cannot normalize.');
  }

  const normalized = {};
  for (const [groupName, values] of Object.entries(concentrations)) {
    normalized[groupName] = values.map((v) => (v / refMean) * 100);
  }

  return normalized;
}

/**
 * Calculate mean and SD for each group.
 * @returns {{ group: string, mean: number, sd: number, n: number }[]}
 */
export function calculateStatistics(normalized) {
  const stats = [];
  for (const [groupName, values] of Object.entries(normalized)) {
    const n = values.length;
    const mean = values.reduce((a, b) => a + b, 0) / n;
    let sd = 0;
    if (n > 1) {
      const sumSqDiff = values.reduce((acc, v) => acc + (v - mean) ** 2, 0);
      sd = Math.sqrt(sumSqDiff / (n - 1)); // Bessel's correction
    }
    stats.push({ group: groupName, mean, sd, n });
  }
  return stats;
}

/**
 * Run independent t-tests (Welch's) comparing each group against the comparison group.
 * @returns {{ [groupName]: { tStat: number|null, pValue: number|null, significance: string } }}
 */
export function runTTests(normalized, comparisonGroup) {
  const refValues = normalized[comparisonGroup];
  if (!refValues) {
    throw new Error(`Comparison group '${comparisonGroup}' not found.`);
  }

  const results = {};
  for (const [groupName, values] of Object.entries(normalized)) {
    if (groupName === comparisonGroup) {
      results[groupName] = { tStat: null, pValue: null, significance: '-' };
      continue;
    }

    if (values.length < 2 || refValues.length < 2) {
      results[groupName] = { tStat: null, pValue: null, significance: 'n/a' };
      continue;
    }

    // Welch's t-test
    const n1 = values.length;
    const n2 = refValues.length;
    const mean1 = values.reduce((a, b) => a + b, 0) / n1;
    const mean2 = refValues.reduce((a, b) => a + b, 0) / n2;
    const var1 = values.reduce((acc, v) => acc + (v - mean1) ** 2, 0) / (n1 - 1);
    const var2 = refValues.reduce((acc, v) => acc + (v - mean2) ** 2, 0) / (n2 - 1);

    const se = Math.sqrt(var1 / n1 + var2 / n2);
    if (se === 0) {
      results[groupName] = { tStat: 0, pValue: 1, significance: 'n.s.' };
      continue;
    }

    const tStat = (mean1 - mean2) / se;

    // Welch-Satterthwaite degrees of freedom
    const num = (var1 / n1 + var2 / n2) ** 2;
    const denom =
      (var1 / n1) ** 2 / (n1 - 1) + (var2 / n2) ** 2 / (n2 - 1);
    const df = num / denom;

    // Two-tailed p-value using jStat t-distribution
    const pValue = 2 * (1 - jStat.studentt.cdf(Math.abs(tStat), df));

    results[groupName] = {
      tStat,
      pValue,
      significance: significanceMarker(pValue),
    };
  }

  return results;
}

function significanceMarker(p) {
  if (p < 0.001) return '***';
  if (p < 0.01) return '**';
  if (p < 0.05) return '*';
  return 'n.s.';
}

File: utils/analysis.py
"""Melanin content analysis: concentration, normalization, statistics, t-tests."""

import numpy as np
import pandas as pd
from scipy import stats


def absorbance_to_concentration(
    absorbance: float, slope: float, intercept: float
) -> float:
    """Convert absorbance to melanin concentration using standard curve.

    Standard curve: Abs = slope * Conc + intercept
    Therefore:      Conc = (Abs - intercept) / slope
    """
    if slope == 0:
        raise ValueError("Slope cannot be zero.")
    return (absorbance - intercept) / slope


def calculate_group_concentrations(
    parsed_data: dict[str, dict[str, float]],
    groups: dict[str, list[str]],
    slope: float,
    intercept: float,
) -> dict[str, list[float]]:
    """Calculate melanin concentrations for each group across all measurements.

    For each measurement file, for each group, averages the absorbance of
    all wells assigned to that group, then converts to concentration.
    Returns {group_name: [concentration_per_measurement]}.
    """
    result: dict[str, list[float]] = {name: [] for name in groups}

    for file_data in parsed_data.values():
        for group_name, wells in groups.items():
            abs_values = [file_data[w] for w in wells if w in file_data]
            if abs_values:
                mean_abs = np.mean(abs_values)
                conc = absorbance_to_concentration(mean_abs, slope, intercept)
                result[group_name].append(conc)

    return result


def normalize_to_reference(
    concentrations: dict[str, list[float]],
    reference_group: str,
) -> dict[str, list[float]]:
    """Normalize all concentrations to the mean of the reference group.

    Each individual value becomes (value / reference_mean) * 100.
    """
    if reference_group not in concentrations:
        raise ValueError(f"Reference group '{reference_group}' not found.")

    ref_values = concentrations[reference_group]
    if not ref_values:
        raise ValueError(f"Reference group '{reference_group}' has no data.")

    ref_mean = np.mean(ref_values)
    if ref_mean == 0:
        raise ValueError("Reference group mean is zero; cannot normalize.")

    normalized: dict[str, list[float]] = {}
    for group_name, values in concentrations.items():
        normalized[group_name] = [(v / ref_mean) * 100 for v in values]

    return normalized


def calculate_statistics(
    normalized: dict[str, list[float]],
) -> pd.DataFrame:
    """Calculate mean and SD for each group.

    Returns DataFrame with columns: Group, Mean, SD, N.
    """
    rows = []
    for group_name, values in normalized.items():
        arr = np.array(values)
        rows.append(
            {
                "Group": group_name,
                "Mean": np.mean(arr),
                "SD": np.std(arr, ddof=1) if len(arr) > 1 else 0.0,
                "N": len(arr),
            }
        )
    return pd.DataFrame(rows)


def run_ttests(
    normalized: dict[str, list[float]],
    comparison_group: str,
) -> dict[str, dict]:
    """Run independent t-tests comparing each group against the comparison group.

    Returns {group_name: {"t_stat": float, "p_value": float, "significance": str}}.
    """
    if comparison_group not in normalized:
        raise ValueError(f"Comparison group '{comparison_group}' not found.")

    ref_values = np.array(normalized[comparison_group])
    results: dict[str, dict] = {}

    for group_name, values in normalized.items():
        if group_name == comparison_group:
            results[group_name] = {
                "t_stat": None,
                "p_value": None,
                "significance": "-",
            }
            continue

        arr = np.array(values)
        if len(arr) < 2 or len(ref_values) < 2:
            results[group_name] = {
                "t_stat": None,
                "p_value": None,
                "significance": "n/a",
            }
            continue

        t_stat, p_value = stats.ttest_ind(arr, ref_values, equal_var=False)
        sig = _significance_marker(p_value)
        results[group_name] = {
            "t_stat": float(t_stat),
            "p_value": float(p_value),
            "significance": sig,
        }

    return results


def _significance_marker(p: float) -> str:
    if p < 0.001:
        return "***"
    elif p < 0.01:
        return "**"
    elif p < 0.05:
        return "*"
    else:
        return "n.s."


def build_results_dataframe(
    parsed_data: dict[str, dict[str, float]],
    groups: dict[str, list[str]],
    concentrations: dict[str, list[float]],
    normalized: dict[str, list[float]],
    stats_df: pd.DataFrame,
    ttest_results: dict[str, dict],
) -> pd.DataFrame:
    """Build a comprehensive results table for display and export."""
    group_names = list(groups.keys())
    filenames = list(parsed_data.keys())

    rows = []
    # Absorbance values per measurement
    for i, fname in enumerate(filenames):
        row = {"Measurement": f"M{i+1} ({fname})"}
        for gname in group_names:
            wells = groups[gname]
            abs_vals = [parsed_data[fname].get(w) for w in wells if w in parsed_data[fname]]
            row[f"{gname} (Abs)"] = np.mean(abs_vals) if abs_vals else None
        rows.append(row)

    # Concentration values
    for i in range(len(filenames)):
        row = {"Measurement": f"M{i+1} Conc."}
        for gname in group_names:
            vals = concentrations.get(gname, [])
            row[f"{gname} (Abs)"] = vals[i] if i < len(vals) else None
        rows.append(row)

    # Normalized values
    for i in range(len(filenames)):
        row = {"Measurement": f"M{i+1} Norm (%)"}
        for gname in group_names:
            vals = normalized.get(gname, [])
            row[f"{gname} (Abs)"] = vals[i] if i < len(vals) else None
        rows.append(row)

    return pd.DataFrame(rows)
